File: algorithm_problems/leetcode/001-050/003. Longest_Substring_Without_Repeating_Characters.js
// https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

var lengthOfLongestSubstring = function(s) {
  let max = 0;
  let visited = new Set();
  let queue = [];
  let count = 0;
  for (let i = 0; i < s.length; i++) {
    if (!queue.length || !visited.has(s[i])) {
      queue.push(s[i]);
      visited.add(s[i]);
      count++;
      if (count > max) max = count;
    } else if (visited.has(s[i])) {
      while (visited.has(s[i])) {
        let first = queue.shift();
        visited.delete(first);
        count--;
      }
      queue.push(s[i]);
      visited.add(s[i]);
      count++;
    }
  }
  return max;
};

console.log(lengthOfLongestSubstring("abcabcbb"));
console.log(lengthOfLongestSubstring("bbbbb"));
console.log(lengthOfLongestSubstring("pwwkew"));
console.log(lengthOfLongestSubstring("aab"));

File: algorithm_problems/leetcode/001-050/003. Longest_Substring_Without_Repeating_Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        max = 0
        visited = set()
        queue = []
        count = 0
        for char in s:
            if not s or char not in visited:
                queue.append(char)
                visited.add(char)
                count += 1
                if count > max:
                    max = count
            elif char in visited:
                while char in visited:
                    first = queue.pop(0)
                    visited.remove(first)
                    count -= 1
                queue.append(char)
                visited.add(char)
                count += 1
                if count > max:
                    max = count

        return max

